ui/src/app/services/api-definition-file.service.ts

```diff
@@ -0,0 +1,171 @@
+            description: 'OpenAPI Spec File',
+            accept: {
+                "application/json": [".json"],
+                "application/x-yaml": [".yaml", ".yml"]
+            }
+        }
+    ];
+
+    public static getFileFormat(file: File): FileFormat {
+        const isJson = file.type === "application/json" || file.name.endsWith(".json");
+        const isYaml = file.name.endsWith(".yaml") || file.name.endsWith(".yml");
+        return isJson ? "json" : isYaml ? "yaml" : undefined;
+    }
+
+    private fileHandle: FileSystemFileHandle;
+
+    constructor(private windowRef: WindowRef, private downloader: DownloaderService) { }
+
+    public get fileSystemAccessApiAvailable() {
+        return 'showOpenFilePicker' in this.windowRef.window;
+    }
+
+    public get hasFileHandle() {
+        return Boolean(this.fileHandle);
+    }
+
+    private async readFile(file: File): Promise<string> {
+        // This is a new API which is not yet available on Safari.
+        // See: https://developer.mozilla.org/en-US/docs/Web/API/Blob/text#Browser_compatibility
+        if ('text' in (file as any)) {
+            return await (file as any).text();
+        }
+
+        return new Promise((resolve, reject) => {
+            const reader: FileReader = new FileReader();
+
+            reader.onload = (fileLoadedEvent) => {
+                resolve((fileLoadedEvent.target as FileReader).result.toString());
+            };
+
+            reader.onerror = (e) => {
+                console.error("Error reading file: ", e);
+                throw new Error("Error reading file.");
+            };
+
+            reader.readAsBinaryString(file);
+        });
+    }
+
+    public async load(file?: File | FileSystemFileHandle): Promise<any> {
+        // We cannot just use `file instanceof FileSystemFileHandle` because it will throw an
+        // exception if that type is not defined (in browsers that do not support the feature)
+        if (this.fileSystemAccessApiAvailable && file instanceof FileSystemFileHandle) {
+            this.fileHandle = file;
+            file = await this.fileHandle.getFile();
+        }
+
+        if (!file) {
+            if (!this.fileSystemAccessApiAvailable) {
+                throw new Error('When file system access API is unavailable, a file must be specified.');
+            }
+
+            let fileHandles: FileSystemFileHandle[];
+            try {
+                fileHandles = await this.windowRef.window.showOpenFilePicker({
+                    multiple: false,
+                    types: ApiDefinitionFileService.filePickerAcceptTypes
+                });
+            } catch (e) {
+                throw new Error("No file selected.");
+            }
+
+            this.fileHandle = fileHandles[0];
+            file = await this.fileHandle.getFile();
+        }
+
+        const contents = await this.readFile(file as File);
+
+        let spec: any;
+        try {
+            spec = parseYaml(contents);
```

Comment:
Previously there were checks to see if the data was in JSON or YAML then using `JSON.parse` or `parseYaml`, but this is actually not necessary since JSON is a subset of YAML and therefore `parseYaml` works for both cases.

Reply:
TIL